Frameworks/Foundation/NSOrderedSet.mm

```diff
@@ -14,311 +14,403 @@
+- (NSUInteger)count {
+    return setContainer.size();
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-- (instancetype)initWithObjects:(id)firstObj {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (NSArray*)array {
+    NSMutableArray* array = [[[NSMutableArray alloc] init] autorelease];
+    auto object = setContainer.data();
+    NSUInteger length = setContainer.size();
+
+    for (NSUInteger i = 0; i < length; i++) {
+        [array addObject:object[i]];
+    }
+
+    return [[array copy] autorelease];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-- (instancetype)initWithObjects:(id _Nonnull const[])objects count:(NSUInteger)cnt {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (NSSet*)set {
+    NSMutableSet* set = [NSMutableSet setWithCapacity:[self count]];
+
+    for (NSUInteger i = 0; i < [self count]; i++) {
+        [set addObject:[self objectAtIndex:i]];
+    }
+
+    return [[set copy] autorelease];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
+*/
+- (instancetype)initWithObject:(id)object {
+    if (object == nil) {
+        [NSException raise:NSInvalidArgumentException format:@"object can not be nil."];
+    }
+
+    id ret[] = { object };
+    return [self initWithObjects:ret count:1];
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithObjects:(id)firstObj, ... {
+    va_list argList;
+    va_start(argList, firstObj);
+    std::vector<id> flatArgs = ConvertVAListToVector((id)firstObj, argList);
+    va_end(argList);
+    return [self initWithObjects:flatArgs.data() count:flatArgs.size()];
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithObjects:(id _Nonnull const[])objects count:(NSUInteger)count {
+    if (objects == nil) {
+        [NSException raise:NSInvalidArgumentException format:@"objects can not be nil."];
+    }
+
+    if (self = [super init]) {
+        for (NSUInteger i = 0; i < count; ++i) {
+            [self _insertObject:objects[i]];
+        }
+    }
+    return self;
+}
+/**
+ @Status Interoperable
 */
 - (instancetype)initWithOrderedSet:(NSOrderedSet*)set {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [self initWithOrderedSet:set range:NSMakeRange(0, [set count]) copyItems:NO];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (instancetype)initWithOrderedSet:(NSOrderedSet*)set copyItems:(BOOL)flag {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [self initWithOrderedSet:set range:NSMakeRange(0, [set count]) copyItems:flag];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (instancetype)initWithOrderedSet:(NSOrderedSet*)orderedSet range:(NSRange)range copyItems:(BOOL)flag {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (orderedSet == nil) {
+        return [self init];
+    }
+
+    auto values = orderedSet->setContainer;
+
+    NSMutableArray* arrayObjs = [NSMutableArray array];
+
+    for (id item : values) {
+        [arrayObjs addObject:item];
+    }
```

Comment:
or, better yet, take bviglietta's suggestion below to delegate to a common NSEnumerator initializer, I'm not fond of making this intermediate array either

---
In reply to: [62376601](https://github.com/Microsoft/WinObjC/pull/482#discussion_r62376601) [](ancestors = 62376601)